hstream/src/HStream/Server/Core/Subscription.hs

```diff
@@ -717,6 +720,16 @@ sendRecords ServerContext{..} subState subCtx@SubscribeContext {..} = do
               resendBatch = mkBatchedRecord batchedRecordCompressionType batchedRecordPublishTime (fromIntegral $ V.length records) records
               resendRecords = ReceivedRecord ids (Just resendBatch)
           void $ sendReceivedRecords logId batchId resendRecordIds resendRecords True
+     where
+      readLoop reader n
+        | n == 0 = return []
+        | otherwise = do
+          res <- S.readerRead reader 1
+          if null res
+             then do
+               Log.warning $ "reader read got empty result, logId " <> Log.build logId <> ", batchId " <> Log.build batchId <> ", retry."
+               readLoop reader (n - 1)
```

Comment:
Also sleep a few seconds before retrying?

Reply:
I don't think it's necessary.
  1. we already set a 10 ms read timeout for a reader
  2. we only retry 3 times, avoiding busy loop